test/cpp_api_parity/functional_impl_check.py

```diff
@@ -0,0 +1,258 @@
+
+        functional_name = compute_functional_name(test_params_dict)
+
+        assert hasattr(torch.nn.functional, functional_name), \
+            "`torch.nn.functional` doesn't have function `{}`. (Discovered while processing\n{}.)".format(
+                functional_name, pprint.pformat(test_params_dict))
+
+        functional_full_name = 'F::' + functional_name
+
+        assert functional_full_name in parity_table['torch::nn::functional'], (
+            "Please add `{}` entry to `torch::nn::functional` section of `test/cpp_api_parity/parity-tracker.md`. "
+            "(Discovered while processing\n{}.)").format(functional_full_name, pprint.pformat(test_params_dict))
+
+        for device in devices:
+            test_params = process_test_params_for_functional(
+                test_params_dict=test_params_dict,
+                device=device,
+                test_instance_class=test_instance_class,
+            )
+            test_name = 'test_torch_nn_functional_{}'.format(test_params.functional_variant_name)
+            unit_test_class.functional_test_params_map[test_name] = test_params
+
+            def test_fn(self):
+                test_torch_nn_functional_variant(
+                    unit_test_class=self, test_params=unit_test_class.functional_test_params_map[self._testMethodName])
+
+            test_fn = decorate_test_fn(
+                test_fn=test_fn,
+                test_cpp_api_parity=test_params_dict.get('test_cpp_api_parity', True),
+                test_cuda=test_params_dict.get('test_cuda', True),
+                has_impl_parity=parity_table['torch::nn::functional'][functional_full_name][0] and
+                test_params_dict.get('has_parity', True),
+                device=device)
+
+            add_test(unit_test_class, test_name, test_fn)
+
+def write_tests_to_test_class(unit_test_class, test_params_dicts, test_instance_class, parity_table, devices):
+    add_torch_nn_functional_impl_parity_tests(
+        parity_table=parity_table,
+        unit_test_class=unit_test_class,
+        test_params_dicts=test_params_dicts,
+        test_instance_class=test_instance_class,
+        devices=devices)
+
+def generate_test_cpp_sources(test_params, template):
+    cpp_args_construction_stmts, _ = compute_cpp_args_construction_stmts_and_forward_arg_symbols(test_params)
+
+    test_cpp_sources = template.substitute(
+        functional_variant_name=test_params.functional_variant_name,
+        cpp_args_construction_stmts=";\n  ".join(cpp_args_construction_stmts),
+        cpp_function_call=test_params.cpp_function_call,
+    )
+    return test_cpp_sources
+
+# Build all C++ tests together, instead of once per test.
+def build_cpp_tests(unit_test_class, print_cpp_source=False):
+    assert len(unit_test_class.functional_test_params_map) > 0
+    cpp_sources = TORCH_NN_COMMON_TEST_HARNESS
+    functions = []
+    functionals_added_cpp_sources = set()
```

Comment:
Why is this a set?

Reply:
One functional can have multiple test dicts, and if we always add the functional's cpp source for each test dict, there would be redefinition issues when we are adding the 2nd test dict / 3rd test dict / etc., and that's why we want to use a set here to keep track of the functionals we already added cpp sources for

Reply:
Oh, I see. This is only done for `sample_functional`, right? If so it would be nice if we could separate out the logic (the sample functional tests vs general tests) because it's a bit confusing what is going on here

Reply:
I am thinking something like the following:
```
functionals_added_cpp_sources = [SAMPLE_FUNCTIONAL_CPP_SOURCE]
for test_name, test_params in unit_test_class.functional_test_params_map.items():
    cpp_sources += generate_test_cpp_sources(test_params=test_params, template=TORCH_NN_FUNCTIONAL_TEST_FORWARD)
    functions.append('{}_test_forward'.format(test_params.functional_variant_name))
    if print_cpp_source:
        print(cpp_sources)
```
This means that we don't need `cpp_api_parity.functional_cpp_sources['sample_functional'] = SAMPLE_FUNCTIONAL_CPP_SOURCE` anymore in `sample_functional.py`, and furthemore we don't need `cpp_api_parity.functional_cpp_sources`.

Reply:
Thanks for the suggestion! Improved accordingly :D